pkg/util/tags_builder.go

```diff
@@ -44,17 +61,171 @@ func (tb *TagsBuilder) Reset() {
+	return &HashingTagsBuilder{
+		hashedTags: newHashedTagsWithCapacity(128),
+	}
+}
+
+// NewHashingTagsBuilderFromSlice return a new TagsBuilder with the input slice for
+// it's internal buffer.
+func NewHashingTagsBuilderFromSlice(tags []string) *HashingTagsBuilder {
+	return &HashingTagsBuilder{newHashedTagsFromSlice(tags)}
+}
+
+// Append appends tags to the builder
+func (tb *HashingTagsBuilder) Append(tags ...string) {
+	for _, t := range tags {
+		tb.data = append(tb.data, t)
+		tb.hash = append(tb.hash, murmur3.StringSum64(t))
+	}
+}
+
+// AppendHashed appends tags and corresponding hashes to the builder
+func (tb *HashingTagsBuilder) AppendHashed(src *HashedTags) {
+	tb.data = append(tb.data, src.data...)
+	tb.hash = append(tb.hash, src.hash...)
+}
+
+// SortUniq sorts and remove duplicate in place
+func (tb *HashingTagsBuilder) SortUniq() {
+	if tb.Len() < 2 {
+		return
+	}
+
+	sort.Sort(tb)
+
+	j := 0
+	for i := 1; i < len(tb.data); i++ {
+		if tb.hash[i] == tb.hash[j] && tb.data[i] == tb.data[j] {
+			continue
+		}
+		j++
+		tb.data[j] = tb.data[i]
+		tb.hash[j] = tb.hash[i]
+	}
+
+	tb.Truncate(j + 1)
+}
+
+// Reset resets the size of the builder to 0 without discaring the internal
+// buffer
+func (tb *HashingTagsBuilder) Reset() {
+	// we keep the internal buffer but reset size
+	tb.data = tb.data[0:0]
+	tb.hash = tb.hash[0:0]
+}
+
 // Truncate retains first n tags in the buffer without discarding the internal buffer
-func (tb *TagsBuilder) Truncate(len int) {
+func (tb *HashingTagsBuilder) Truncate(len int) {
 	tb.data = tb.data[0:len]
+	tb.hash = tb.hash[0:len]
 }
 
 // Get returns the internal slice
-func (tb *TagsBuilder) Get() []string {
+func (tb *HashingTagsBuilder) Get() []string {
 	return tb.data
 }
 
+// Hashes returns the internal slice of tag hashes
+func (tb *HashingTagsBuilder) Hashes() []uint64 {
+	return tb.hash
+}
+
 // Copy makes a copy of the internal slice
-func (tb *TagsBuilder) Copy() []string {
+func (tb *HashingTagsBuilder) Copy() []string {
 	return append(make([]string, 0, len(tb.data)), tb.data...)
 }
+
+// Less implements sort.Interface.Less
+func (tb *HashingTagsBuilder) Less(i, j int) bool {
+	// FIXME(vickenty): could sort using hashes, which is faster, but a lot of tests check for order.
+	return tb.data[i] < tb.data[j]
+}
+
+// Swap implements sort.Interface.Swap
+func (tb *HashingTagsBuilder) Swap(i, j int) {
+	tb.hash[i], tb.hash[j] = tb.hash[j], tb.hash[i]
+	tb.data[i], tb.data[j] = tb.data[j], tb.data[i]
+}
+
+// Len implements sort.Interface.Len
+func (tb *HashingTagsBuilder) Len() int {
+	return len(tb.data)
+}
+
+// HashedTags is an immutable slice of pre-hashed tags.
+type HashedTags struct {
+	hashedTags
+}
+
+// NewHashedTagsFromSlice creates a new instance, re-using tags as the internal slice.
+func NewHashedTagsFromSlice(tags []string) HashedTags {
+	return HashedTags{newHashedTagsFromSlice(tags)}
+}
+
+// Slice returns a shared sub-slice of tags from t.
```

Comment:
This method exists so we can read and modify the slice in ckey.Generate. Same is partially true for `Get()` as well.

Reply:
> It should not be modified.
It can not be (directly), HashedTags has no mutators. 

It can still be invalidated (by mistake) by mutating slices returned from `Get` or `Hashes`, so not sure we can promise it to remain valid either. :/